### PR TITLE
JENKINS-23896 Give CommandInterpreter.join(Proc p) access to Build

### PR DESCRIPTION
**Pull request by ringerc**
The CommandInterpreter.join(Proc p) method is supposed to be an extension
point to allow an UNSTABLE build result to be set, but it has no access to
the Build object with which to do so.

Add a getBuild() method that returns the build passed to perform(...). A better
API would be a new join(...) with more arguments, but that'd break out of tree
plugins that might already be using this for other purposes.

### CODE REVIEW

**Comment by olenz** [comment]
Why isn't this merged yet? Are there any blockers?

File: core/src/main/java/hudson/tasks/CommandInterpreter.java [hunk 1]
```diff
@@ -99,6 +118,7 @@ public boolean perform(AbstractBuild<?,?> build, Launcher launcher, TaskListener
                 Util.displayIOException(e, listener);
                 e.printStackTrace(listener.fatalError(Messages.CommandInterpreter_CommandFailed()));
             }
+            this.build = null;
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tasks/CommandInterpreter.java]: And <code>transient</code>... I doubt such approach will work with <code>SimpleBuilder</code>s

**Comment by oleg-nenashev** [comment]
The change is incomplete.
I have a weak :-1: due to <code>SimpleBuildStep</code> integration issues (and other stuff, which may be potentially used from multiple threads)

**Comment by jglick** [comment]
You can already use

```java
Executor e = Executor.currentExecutor();
if (e != null) {
    Queue.Executable x = e.getCurrentExecutable();
    if (x instanceof Run) {
        ((Run<?,?>) x).setResult(Result.UNSTABLE);
    }
}
```

This will not work from a `SimpleBuildStep`, but neither would the proposed trick here.

Or you could just override `perform` to set a `ThreadLocal<Run<?,?>>` with your build, without any core change. That could be done in a `SimpleBuildStep`, too.

If we do want a core change for convenience, there is no need for any field and getter. Simply introduce a new overload of `join` that takes `Run<?,?>`, call it from `perform` with a local variable, and put in a compatibility shim for overrides of the old and now deprecated overload.

:-1:

**Pull request closed without merging**